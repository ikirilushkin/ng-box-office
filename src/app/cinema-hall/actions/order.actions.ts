import { createAction, props } from '@ngrx/store';

export const proceedOrder = createAction(
  '[Cinema Hall] Proceed Order',
  props<{ ids: number[] }>()
);

export const cancelOrder = createAction(
  '[Cinema Hall] Cancel Order',
  props<{ ids: number[] }>()
);
