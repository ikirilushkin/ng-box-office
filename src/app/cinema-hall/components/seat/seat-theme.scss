@import '~@angular/material/theming';

@mixin seat-theme($theme) {
  $primary: map-get($theme, primary);
  $accent: map-get($theme, accent);
  $warn: map-get($theme, warn);

  .seat {
    background-color: mat-color($primary);

    &.booked {
      background-color: mat-color($accent);
    }

    &.ordered {
      background-color: mat-color($warn);
    }
  }
}
