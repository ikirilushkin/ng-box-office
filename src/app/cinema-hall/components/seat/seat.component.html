<span class="seat" (click)="toggle()"
    [ngClass]="{ordered: seat.status === 'ORDERED', booked: seat.status === 'BOOKED'}">
    {{ seat.seatNumber }}
</span>