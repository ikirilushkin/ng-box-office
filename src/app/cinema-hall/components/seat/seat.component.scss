.seat {
  width: 30px;
  height: 30px;
  display: inline-block;
  border: 1px solid;
  cursor: pointer;
  text-align: center;
  border-radius: 4px;
  border-color: green;

  &.booked {
    border-color: yellow;
  }

  &.ordered {
    border-color: red;
  }
}
