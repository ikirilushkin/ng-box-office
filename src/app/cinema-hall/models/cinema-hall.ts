export type SeatStatus = 'ORDERED' | 'BOOKED' | 'FREE';

export interface Seat {
  id: number;
  row: number;
  seatNumber: number;
  status: SeatStatus;
}

export interface CinemaRow {
  rowNumber: number;
  seats: Seat[];
}

export interface CinemaHallView {
  rows: CinemaRow[];
}

export function generateMockSeat(status: SeatStatus): Seat {
  return {
    id: 1000,
    row: 1,
    seatNumber: 1,
    status
  };
}
