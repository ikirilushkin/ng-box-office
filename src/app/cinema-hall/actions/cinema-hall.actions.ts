import { Seat } from '@bo/cinema-hall/models';
import { createAction, props } from '@ngrx/store';

export const loadCinemaHallSeats = createAction('[Cinema Hall] Load Seats');

export const loadCinemaHallSeatsSuccess = createAction(
  '[Cinema Hall] Load Seats Success',
  props<{ seats: Seat[] }>()
);

export const bookSeat = createAction(
  '[Cinema Hall] Book Seat',
  props<{ seatId: number }>()
);

export const cancelSeatBooking = createAction(
  '[Cinema Hall] Cancel Seat Booking',
  props<{ seatId: number }>()
);

export const proceedOrder = createAction(
  '[Cinema Hall] Proceed Order',
  props<{ ids: number[] }>()
);

export const cancelOrder = createAction(
  '[Cinema Hall] Cancel Order',
  props<{ ids: number[] }>()
);
