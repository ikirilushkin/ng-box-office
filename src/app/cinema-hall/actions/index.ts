import * as CinemaHallActions from './cinema-hall.actions';

export { CinemaHallActions };
