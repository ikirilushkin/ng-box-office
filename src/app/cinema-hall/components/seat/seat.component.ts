import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  Output
} from '@angular/core';
import { Seat } from '@bo/cinema-hall/models';

@Component({
  selector: 'bo-seat',
  templateUrl: './seat.component.html',
  styleUrls: ['./seat.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class SeatComponent {
  @Input() seat: Seat;

  @Output() toggleSeat: EventEmitter<Seat> = new EventEmitter();

  toggle() {
    if (this.seat.status !== 'ORDERED') {
      this.toggleSeat.emit(this.seat);
    }
  }
}
