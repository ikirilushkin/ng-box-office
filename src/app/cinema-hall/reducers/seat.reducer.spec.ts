import * as fromSeat from '@bo/cinema-hall/reducers/seat.reducer';
import { reducer } from '@bo/cinema-hall/reducers/seat.reducer';
import { OrderActions, SeatActions } from '../actions';
import { generateMockSeat } from '../models';

describe('SeatReducer', () => {
  describe('undefined action', () => {
    it('should return default state', () => {
      const result = reducer(undefined, {} as any);
      expect(result).toMatchSnapshot();
    });
  });

  describe('seat actions', () => {
    const initialState: fromSeat.State = {
      ids: [],
      entities: {}
    };

    it('should load cinema hall seats', () => {
      const seat1 = generateMockSeat('FREE');
      const seat2 = { ...seat1, id: 1001 };
      const action = SeatActions.loadCinemaHallSeatsSuccess({
        seats: [seat1, seat2]
      });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });

    it('should book seat', () => {
      const seat1 = generateMockSeat('FREE');
      const seat2 = { ...seat1, id: 1001 };
      const initialState1: fromSeat.State = {
        ids: [seat1.id, seat2.id],
        entities: { [seat1.id]: seat1, [seat2.id]: seat2 }
      };
      const action = SeatActions.bookSeat({ seatId: seat2.id });
      const result = reducer(initialState1, action);
      expect(result).toMatchSnapshot();
    });

    it('should cancel seat booking', () => {
      const seat1 = generateMockSeat('FREE');
      const seat2 = { ...generateMockSeat('BOOKED'), id: 1001 };
      const initialState1: fromSeat.State = {
        ids: [seat1.id, seat2.id],
        entities: { [seat1.id]: seat1, [seat2.id]: seat2 }
      };
      const action = SeatActions.cancelSeatBooking({ seatId: seat2.id });
      const result = reducer(initialState1, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('order actions', () => {
    it('should proceed order', () => {
      const seat1 = generateMockSeat('FREE');
      const seat2 = { ...seat1, id: 1001 };
      const initialState1: fromSeat.State = {
        ids: [seat1.id, seat2.id],
        entities: { [seat1.id]: seat1, [seat2.id]: seat2 }
      };
      const action = OrderActions.proceedOrder({ ids: [seat1.id, seat2.id] });
      const result = reducer(initialState1, action);
      expect(result).toMatchSnapshot();
    });

    it('should cancel order', () => {
      const seat1 = generateMockSeat('BOOKED');
      const seat2 = { ...seat1, id: 1001 };
      const initialState1: fromSeat.State = {
        ids: [seat1.id, seat2.id],
        entities: { [seat1.id]: seat1, [seat2.id]: seat2 }
      };
      const action = OrderActions.cancelOrder({ ids: [seat1.id, seat2.id] });
      const result = reducer(initialState1, action);
      expect(result).toMatchSnapshot();
    });
  });
});
