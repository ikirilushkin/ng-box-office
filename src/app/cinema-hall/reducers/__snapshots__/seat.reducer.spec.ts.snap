// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeatReducer order actions should cancel order 1`] = `
Object {
  "entities": Object {
    "1000": Object {
      "id": 1000,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
    "1001": Object {
      "id": 1001,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
  },
  "ids": Array [
    1000,
    1001,
  ],
}
`;

exports[`SeatReducer order actions should proceed order 1`] = `
Object {
  "entities": Object {
    "1000": Object {
      "id": 1000,
      "row": 1,
      "seatNumber": 1,
      "status": "ORDERED",
    },
    "1001": Object {
      "id": 1001,
      "row": 1,
      "seatNumber": 1,
      "status": "ORDERED",
    },
  },
  "ids": Array [
    1000,
    1001,
  ],
}
`;

exports[`SeatReducer seat actions should book seat 1`] = `
Object {
  "entities": Object {
    "1000": Object {
      "id": 1000,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
    "1001": Object {
      "id": 1001,
      "row": 1,
      "seatNumber": 1,
      "status": "BOOKED",
    },
  },
  "ids": Array [
    1000,
    1001,
  ],
}
`;

exports[`SeatReducer seat actions should cancel seat booking 1`] = `
Object {
  "entities": Object {
    "1000": Object {
      "id": 1000,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
    "1001": Object {
      "id": 1001,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
  },
  "ids": Array [
    1000,
    1001,
  ],
}
`;

exports[`SeatReducer seat actions should load cinema hall seats 1`] = `
Object {
  "entities": Object {
    "1000": Object {
      "id": 1000,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
    "1001": Object {
      "id": 1001,
      "row": 1,
      "seatNumber": 1,
      "status": "FREE",
    },
  },
  "ids": Array [
    1000,
    1001,
  ],
}
`;

exports[`SeatReducer undefined action should return default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
}
`;
